File: mergeSort.c
#include <stdio.h>
#include <stdlib.h>
#include <windows.h>  // Biblioteca do Windows para alta precisão
#define TAM 50000

void mergeSort(int vet[], int inicio, int fim);
void intercala(int vet[], int inicio, int meio, int fim);

int main() {
    int vet[TAM] = {10, 9, 100, 900, 289};

    LARGE_INTEGER inicio, fim, freq;
    double tempo_execucao;

    QueryPerformanceFrequency(&freq);     // Frequência do contador
    QueryPerformanceCounter(&inicio);     // Tempo inicial

    mergeSort(vet, 0, TAM - 1);

    QueryPerformanceCounter(&fim);        // Tempo final

    tempo_execucao = (double)(fim.QuadPart - inicio.QuadPart) / freq.QuadPart;


    printf("\nTempo de execução: %.9f segundos\n", tempo_execucao);

    return 0;
}

void mergeSort(int vet[], int inicio, int fim) {
    if (inicio < fim) {
        int meio = (inicio + fim) / 2;
        mergeSort(vet, inicio, meio);
        mergeSort(vet, meio + 1, fim);
        intercala(vet, inicio, meio, fim);
    }
}

void intercala(int vet[], int inicio, int meio, int fim) {
    int inicio1 = inicio;
    int inicio2 = meio + 1;
    int posLivre = 0;
    int vetAux[TAM];

    while (inicio1 <= meio && inicio2 <= fim) {
        if (vet[inicio1] <= vet[inicio2]) {
            vetAux[posLivre++] = vet[inicio1++];
        } else {
            vetAux[posLivre++] = vet[inicio2++];
        }
    }

    while (inicio1 <= meio) {
        vetAux[posLivre++] = vet[inicio1++];
    }

    while (inicio2 <= fim) {
        vetAux[posLivre++] = vet[inicio2++];
    }

    for (int i = 0; i < posLivre; i++) {
        vet[inicio + i] = vetAux[i];
    }
}

File: MergeSort.py
import time

def intercala(inicio, meio, fim, X):
    aux = [0] * (fim - inicio + 1)
    inicio_v01 = inicio
    inicio_v02 = meio + 1
    poslivre = 0

    while inicio_v01 <= meio and inicio_v02 <= fim:
        if X[inicio_v01] <= X[inicio_v02]:
            aux[poslivre] = X[inicio_v01]
            inicio_v01 += 1
        else:
            aux[poslivre] = X[inicio_v02]
            inicio_v02 += 1
        poslivre += 1

    while inicio_v01 <= meio:
        aux[poslivre] = X[inicio_v01]
        inicio_v01 += 1
        poslivre += 1

    while inicio_v02 <= fim:
        aux[poslivre] = X[inicio_v02]
        inicio_v02 += 1
        poslivre += 1

    for i in range(len(aux)):
        X[inicio + i] = aux[i]

def mergesort(inicio, fim, v):
    if inicio < fim:
        meio = (inicio + fim) // 2
        mergesort(inicio, meio, v)
        mergesort(meio + 1, fim, v)
        intercala(inicio, meio, fim, v)


vetor = []

inicio_tempo = time.time()
mergesort(0, len(vetor) - 1, vetor)
fim_tempo = time.time()

print("Vetor ordenado:")
print(vetor)
print(f"\nTempo de execução: {fim_tempo - inicio_tempo:.6f} segundos")